src/main/scala/is/hail/annotations/StagedRegionValueBuilder.scala

```diff
@@ -127,6 +127,15 @@ class StagedRegionValueBuilder private(val fb: FunctionBuilder[_], val typ: Type
 
   def addStruct(t: TStruct, f: (StagedRegionValueBuilder => Code[Unit]), init: LocalRef[Boolean] = null): Code[Unit] = f(new StagedRegionValueBuilder(fb, t, this))
 
+  def addAnnotation(t: Type): (Code[_]) => Code[Unit] = t.fundamentalType match {
```

Comment:
Hmm, I'm not sure I'd call this `addAnnotation`.  An annotation is an `Annotation = Any`, this is a `Code` with an unknown type (very different than `Code[Any]`).

Reply:
As I understand it:

`RegionValueBuilder.addAnnotation`
 - the value may be missing
 - `advance` is called for each call
 - Booleans are stored as 8-bit values

`StagedRegionValueBuilder.addAnnotation`
 - the value is never missing
 - `advance` is never called
 - Booleans are stored as 32-bit values

On the first point, I guess this should really be called `addPresentAnnotation` and maybe we want an `addAnnotation` that returns a `(Code[Boolean], Code[_]) => Code[Unit]`?

On the second point, I could rewrite `StagedRegionValueBuilder` to match `RegionValueBuilder` if that satisfies the issue.

On the third point, I think this is just a straight up bug. I'll fix with a commit that adds an `addBoolean` method that mimics the behavior of `RegionValueBuilder.addBoolean`.

---

> this is a `Code` with an unknown type (very different than `Code[Any]`)

Maybe this is the real source of my misunderstanding, but that's not what I think of `Code[_]` as. The compiler always knows at compile-time type of the value placed on the stack by `Code[_]`, I just haven't bothered to convince the type system of it.

Is the point you're making that `Any` is necessarily a type containing exclusively boxed values?

Reply:
We just have a confusion about terminology:

In my mind, an annotation is an `Annotation = Any` representing something of some type `t: Type` which is represented as native java objects.

A region value is a `Long` that is an offset into a region representing something of some type `t: Type` which is stored using our custom physical layout.

For example, RegionValueBuilder has `addAnnotation(t: Type, a: Any)` and `addRegionValue(t: Type, rv: RegionValue)`.

Your function is neither of these.  It just adds primitives.  So maybe `addPrimitive(t: Type)(c: Code[_]): Code[Unit]`?

 > addPresentAnnotation

We use missing and defined, not present.

 > Booleans are stored as 32-bit values

How's that?  If true, certainly a bug.

Reply:
Just to be clear, my complaint is the name `addAnnotation`, because annotations are native Java objects.

Reply:
I'll rename to `addPrimitive`. Is our intention not to codify `addAnnotation`? I intended this as a stub for that (I left the non-primitives unimplemented because the current structure of `StagedRegionValueBuilder` didn't match `RegionValueBuilder` enough to make the translation simple and I wanted to focus on finishing the compiler).

Reply:
> Is our intention not to codify `addAnnotation`?

My intention was to add it if necessary as a temporary measure.  I don' think it will be necessary.  Code that makes it all the way to `StagedRVB` won't have legacy annotations.  In the end annotation should only be used for interacting with Spark (e.g. KeyTable.to_dataframe).  Note, `addAnnotation` would take a `Code[Any]`, not a `Code[_]`.  It will really be an any, and if you need e.g. an Int, there will have to be a cast/unboxing in the bytecode.